src/epstats/server/req.py

```diff
@@ -139,28 +138,40 @@ def to_check(self):
         return EvSrmCheck(self.id, self.name, self.denominator)
 
 
-class FilterScope(str, Enum):
-    """
-    Scope of data where to apply the filter.
-    """
-
-    exposure = "exposure"
-    goal = "goal"
-
-
 class Filter(BaseModel):
     """
     Filter specification for data to evaluate.
     """
 
     dimension: str = Field(title="Name of the dimension")
 
-    value: List = Field(title="List of possible values")
+    value: List[Any] = Field(title="List of possible values")
 
     scope: FilterScope = Field(
         title="Scope of the filter", description="Scope of the filter is either `exposure` or `goal`."
     )
 
+    @validator("dimension")
+    def dimension_must_be_not_empty(cls, value):
+        if not value:
+            raise ValueError("we expect dimension to be non-empty")
+        return value
+
+    @validator("scope")
+    def scope_must_be_not_empty(cls, value):
+        if not value:
+            raise ValueError("we expect scope to be either `exposure` or `goal`")
+        return value
+
+    @validator("value")
+    def value_must_be_not_empty(cls, value):
+        if not value:
+            raise ValueError("we expect value to be non-empty")
+        return value
+
+    def to_filter(self):
+        return EvFilter(self.dimension, self.value, self.scope)
+
 
```

Comment:
Since you handle missing "values", is it mandatory for API to fill them? 
And what should be in place of "values" if I want general results (no dimensions)?

Reply:
These validators apply only if there is some `filter` in the query. If we keep request data with no `filter`, these validations are not used and we do not filter any data. See `Optional` in `filter: Optional[List[Filter]] = Field(...)`.